src/main/java/org/voltdb/exportclient/KinesisFirehoseExportClient.java

```diff
@@ -0,0 +1,264 @@
+    }
+
+    class KinesisFirehoseExportDecoder extends ExportDecoderBase {
+        private final ListeningExecutorService m_es;
+
+        private ArrayList<Record> m_records;
+
+        @Override
+        public ListeningExecutorService getExecutor() {
+            return m_es;
+        }
+
+        public KinesisFirehoseExportDecoder(AdvertisedDataSource source)
+        {
+            super(source);
+            m_es = CoreUtils.getListeningSingleThreadExecutor(
+                    "Kinesis Firehose Export decoder for partition " + source.partitionId
+                    + " table " + source.tableName
+                    + " generation " + source.m_generation, CoreUtils.MEDIUM_STACK_SIZE);
+        }
+
+        @Override
+        public boolean processRow(int rowSize, byte[] rowData) throws RestartBlockException
+        {
+            StringWriter stringer = new StringWriter();
+            CSVWriter csv = new CSVWriter(stringer, m_seperator);
+            Record record = new Record();
+            try {
+                final ExportRowData row = decodeRow(rowData);
+                if (!writeRow(row.values, csv, m_skipInternals, m_binaryEncoding, m_ODBCDateformat.get())) {
+                    return false;
+                }
+                csv.flush();
+
+                String data = stringer.toString();
+                record.withData(ByteBuffer.wrap(data.getBytes()));
+            } catch(IOException e) {
+                rateLimitedLogError(m_logger, "Failed to build record: %s", Throwables.getStackTraceAsString(e));
+                throw new RestartBlockException(true);
+            } finally {
+                try { csv.close(); } catch (IOException e) {}
+            }
+
+            m_records.add(record);
+            return true;
+        }
+
+        @Override
+        public void sourceNoLongerAdvertised(AdvertisedDataSource source)
+        {
+            m_es.shutdown();
+            try {
+                m_es.awaitTermination(365, TimeUnit.DAYS);
+            } catch (InterruptedException e) {
+                Throwables.propagate(e);
+            }
+        }
+
+        @Override
+        public void onBlockStart() throws RestartBlockException
+        {
+            m_records = new ArrayList<Record>();
+        }
+
+        @Override
+        public void onBlockCompletion() throws RestartBlockException
+        {
+            try {
+                int recordsSize = m_records.size();
+                List<Record> recordsList;
+                int sleepTime = 0;
+                while (recordsSize > 0) {
+                    if (sleepTime > 0)
+                        Thread.sleep(sleepTime);
+                    // PutRecordBatchRequest can not contain more than 500 records
+                    if (recordsSize > 500) {
+                        recordsList = m_records.subList(0, 499);
+                        m_records = new ArrayList<Record>(m_records.subList(500, recordsSize-1));
```

Comment:
why two slices? just increment an positional tracker by 500 and use it as a basis to new slice